# Pan-genome analysis of all assemblies

## Pan-genome analysis overview:

1. RNA mapping rates

2. K-mer spectrum analysis

3. LTR Assembly Index

4. Collinearity analysis

5. BUSCO evaluation

6. Gene family analysis

   

## RNA mapping rates
To assess the quality of each accession, the 83 public RNA-seq data are downloaded
```shell
# Download data
for i in less SRR.txt \ ; do  echo "/public/home/software/sratoolkit.2.10.8-ubuntu64/bin/prefetch"   "$i" ; done

# Use fastq-dump transfer SRA to FASTQ files
fasterq-dump --split-3 --gzip SRR11111.sra

# Use fastp to clean RNA-seq reads
cat fastq.list | parallel -j 10  'fastp -i ./{}_1.fastq.gz  -o ./{}_1.fastq_clean.gz -I ./{}_2.fastq.gz -O ./{}_2.fastq_clean.gz -j ./{}.json -h ./{}.html'

# Use STAR index genome
STAR --runThreadN 15 --runMode genomeGenerate --genomeDir /vol3/agis/huangsanwen_group/zhangzhiyang/02.transcript/01.STAR_index --genomeSAindexNbases 13 --genomeFastaFiles genome.fa --sjdbGTFfile gene_models.gtf --sjdbOverhang 149

# Use STAR to map reads to genome
cat fastq.list | parallel -j 10  'STAR --runThreadN 10 --genomeDir /vol3/agis/huangsanwen_group/zhangzhiyang/02.transcript/01.STAR_index/ --readFilesCommand gunzip -c --outSAMtype BAM SortedByCoordinate   --outFileNamePrefix {} --outSAMattrIHstart 0 --readFilesIn ./{}_1.fastq_clean.gz  ./{}_2.fastq_clean.gz

# Stats mapping rate
cat sample.list | parallel -j 10 'samtools flagstas {}.bam > {}.stats' 

```



## KAT (The K-mer Analysis Toolkit)

The reference-free quality control (QC) toolkit `KAT` is applied to generate 31-mer spectra plots using HiFi reads. (https://github.com/TGAC/KAT)

````shell
source /public/software/modules-4.5.1/init/bash && export MODULEPATH=/public/home/software/modulefiles:$MODULEPATH
module load KAT/2.4.2
cat sample.id |  parallel -j 64 'sh kat.sh {}'
````



## LTR Assembly Index
To estimate the contiguity of repetitive elements across assemblies, an LTR assembly index (LAI) is calculated using `LTR_retriever`

```shell
for i in `awk '{print $1}' sample.list`;
do
 perl ~/las/git_bin/LTR_retriever/LAI \
    -genome $(echo $i|perl -nle 's/\..*//; print $_')/$i \
    -intact $(echo $i|perl -nle 's/\..*//; print $_')/$i.mod.EDTA.raw/LTR/$i.mod.pass.list \
    -all $i.out.mod \
    -q -iden 94.854 \
    -totLTR 76.34 \
    -t 2 &
done
```



## Collinearity analysis

Comparision collinearity between Heinz 1706 and other assemblies using `minimap2` and visualized by `dotPlotly`

```shell
cat sample.id |  parallel -j 2 'minimap2 -x asm5 -t 8 ./Heinz_1706.fasta {}.fasta > {}.paf
cat sample.id |  parallel -j 2 'minimap2 -x asm5 -t 8 ./LA2093.fasta {}.fasta > {}.paf
cat sample.id |  parallel -j 2 'pafCoordsDotPlotly.R -i {}.paf -o {}.out -s -t -m 500 -q 500000 -k 7 -l'
```

â€‹		

## BUSCO evaluation
The final assemblies are evaluted completeness of collinearity and `leudicots_odb10` is selected as the input parameters for calculating %BUSCO completeness
```shell
source /public/software/modules-4.5.1/init/bash && export MODULEPATH=/public/home/software/modulefiles:$MODULEPATH
module load busco/5.2.2
cat sample.id |  parallel -j 2 'busco -c 30 -i {}.fasta -o {}_busco -m geno -l ./1eudicots_odb10/ --offline'
```



## Gene family analysis

The core and dispensable gene sets among the 46 accessions were generated by the software `Orthodfinder`. The `CD-HIT` is used to remove redundent proteins, core gene families, softcore gene families,  dispensable gene families, private gene families are defined as articles.

```shell
~/miniconda3/envs/annotation/bin//cd-hit-est -c 1 -aS 1 -i cds/1706.cds.fa -o filter/1706.fa -T 8
diamond blastp -d diamondDBSpecies -q Species.fa -o Blast.txt --more-sensitive -p 8 --quiet -e 1e-5 --compress 1
orthofinder -t 64 -a 64 -og -S diamond -b ./filter_pep/OrthoFinder/Results_Jan17/WorkingDirectory

```





